### PR TITLE
Fix Backwards Compatible Parquet List Encodings (#1915)

### LINKED ISSUE #1915: incompatible arrow schema, expected struct got List
**Describe the bug**
```rust
use std::{fs::File, path::Path, sync::Arc};

use arrow::record_batch::RecordBatchReader;
use parquet::{
    arrow::{ArrowReader, ParquetFileArrowReader, ProjectionMask},
    basic::Compression,
    data_type::Int32Type,
    file::{
        properties::{WriterProperties, WriterVersion},
        writer::SerializedFileWriter,
    },
    schema::parser::parse_message_type,
};

const MESSAGE_TYPE: &'static str = "
message Log {
  OPTIONAL INT32 eventType;
  REPEATED INT32 category;
  REPEATED group filter {
    OPTIONAL INT32 error;
  }
}
";

fn main() {
    write();
    read();
}

fn write() {
    let path = Path::new("sample.parquet");
    let file = File::create(&path).unwrap();
    let schema = Arc::new(parse_message_type(MESSAGE_TYPE).unwrap());

    let props = Arc::new(
        WriterProperties::builder()
            .set_compression(Compression::SNAPPY)
            .set_writer_version(WriterVersion::PARQUET_2_0)
            .build(),
    );
    let mut writer = SerializedFileWriter::new(file, schema, props).unwrap();
    let mut row_group_writer = writer.next_row_group().unwrap();

    // column 0
    let mut col_writer = row_group_writer
        .next_column()
        .expect("next column")
        .unwrap();
    col_writer
        .typed::<Int32Type>()
        .write_batch(&[1], Some(&[1]), None)
        .expect("writing column");
    col_writer.close().expect("close column");
    // column 1
    let mut col_writer = row_group_writer
        .next_column()
        .expect("next column")
        .unwrap();
    col_writer
        .typed::<Int32Type>()
        .write_batch(&[1, 1], Some(&[0, 1]), Some(&[0, 1]))
        .expect("writing column");
    col_writer.close().expect("close column");
    // column 2
    let mut col_writer = row_group_writer
        .next_column()
        .expect("next column")
        .unwrap();
    col_writer
        .typed::<Int32Type>()
        .write_batch(&[1], Some(&[1]), Some(&[1]))
        .expect("writing column");
    col_writer.close();

    let rg_md = row_group_writer.close().expect("close row group");
    println!("total rows written: {}", rg_md.num_rows());

    writer.close().unwrap();
}

fn read() {
    let file = File::open("sample.parquet").unwrap();

    let mut arrow_reader = ParquetFileArrowReader::try_new(file).unwrap();
    let mask = ProjectionMask::leaves(arrow_reader.parquet_schema(), [0]);

    println!(
        "Converted arrow schema is: {}",
        arrow_reader.get_schema().unwrap()
    );
    println!(
        "Arrow schema after projection is: {}",
        arrow_reader.get_schema_by_columns(mask.clone()).unwrap()
    ); // error incompatible arrow schema, expected struct got List

    let mut unprojected = arrow_reader.get_record_reader(2048).unwrap();
    println!("Unprojected reader schema: {}", unprojected.schema());

    let mut record_batch_reader = arrow_reader
        .get_record_reader_by_columns(mask, 2048)
        .unwrap();

    for maybe_record_batch in record_batch_reader {
        let record_batch = maybe_record_batch.unwrap();
        if record_batch.num_rows() > 0 {
            println!("Read {} records.", record_batch.num_rows());
        } else {
            println!("End of file!");
        }
    }
}

```

**To Reproduce**

**Expected behavior**

**Additional context**
https://docs.rs/parquet/16.0.0/parquet/arrow/index.html#example-of-reading-parquet-file-into-arrow-record-batch
throw error:
Arrow schema after projection is: Field { name: "eventType", data_type: Int32, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: ArrowError("incompatible arrow schema, expected struct got List(Field { name: \"filter\", data_type: Struct([Field { name: \"error\", data_type: Int32, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }]), nullable: false, dict_id: 0, dict_is_ordered: false, metadata: None })")', example/test.rs:99:64

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #1915

# Rationale for this change
 

As described in https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#backward-compatibility-rules there are various backwards compatibility rules for converting parquet schema to lists. This logic is applied by the schema inference logic added in https://github.com/apache/arrow-rs/pull/1682, which "promotes" bare-repeated fields to be ListArray in the generated arrow schema.

This works well, however, the mechanism by which array_reader is constructed would perform this inference again, using the an inferred arrow schema, as opposed to any schema embedded in the parquet metadata. Aside from being potentially inefficient, this causes issues if the provided arrow schema no longer matches the parquet schema, e.g. a field has been promoted to a different arrow representation.

# What changes are included in this PR?

This pushes projection into the array reader construction, in addition to being more efficient, this avoids the double-inference issues. We infer the schema once, and then project this.

# Are there any user-facing changes?

No

### CODE REVIEW

File: parquet/src/arrow/arrow_reader/mod.rs [hunk 1]
```diff
@@ -2313,4 +2328,66 @@ mod tests {
         assert_ne!(1024, num_rows);
         assert_eq!(reader.batch_size, num_rows as usize);
     }
+
+    #[test]
```
Comment (tustvold) [inline parquet/src/arrow/arrow_reader/mod.rs:2332]: This is based on the test by @liyongjing, thank you for the great reproducer. I had to tweak the levels a little for the generated parquet to be valid, but the spirit remains the same :smile:

File: parquet/src/arrow/arrow_reader/mod.rs [hunk 2]
```diff
@@ -82,15 +84,17 @@ impl<T> ArrowReaderBuilder<T> {
             false => metadata.file_metadata().key_value_metadata(),
         };
 
-        let schema = Arc::new(parquet_to_arrow_schema(
+        let (schema, fields) = parquet_to_array_schema_and_fields(
```
Comment (tustvold) [inline parquet/src/arrow/arrow_reader/mod.rs:87]: Here we can see that we perform the inference logic once, and keep the ParquetField around to later project

File: parquet/src/arrow/arrow_reader/mod.rs [hunk 3]
```diff
@@ -420,9 +432,11 @@ impl<T: ChunkReader + 'static> ArrowReaderBuilder<SyncReader<T>> {
                     break;
                 }
 
-                let projection = predicate.projection().clone();
-                let array_reader =
-                    build_array_reader(Arc::clone(&self.schema), projection, &reader)?;
+                let array_reader = build_array_reader(
```
Comment (tustvold) [inline parquet/src/arrow/arrow_reader/mod.rs:435]: We no longer do potentially expensive schema inference for every filter :tada:

FYI @thinkharderdev

**Comment by tustvold** [comment]
Oops I accidentally included a test parquet file, will remove

File: parquet/src/arrow/array_reader/builder.rs [hunk 4]
```diff
@@ -26,40 +26,43 @@ use crate::arrow::array_reader::{
...
-    mask: ProjectionMask,
+    field: Option<&ParquetField>,
+    mask: &ProjectionMask,
     row_groups: &dyn RowGroupCollection,
 ) -> Result<Box<dyn ArrayReader>> {
-    let field = convert_schema(&row_groups.schema(), mask, Some(arrow_schema.as_ref()))?;
+    let reader = field
+        .and_then(|field| build_reader(field, mask, row_groups).transpose())
+        .transpose()?
+        .unwrap_or_else(|| make_empty_array_reader(row_groups.num_rows()));
```
Comment (thinkharderdev) [inline parquet/src/arrow/array_reader/builder.rs:47]: Not a comment on this PR per se, but I don't understand why we swallow the error here and return and empty array reader

Comment (thinkharderdev) [reply parquet/src/arrow/array_reader/builder.rs:47]: Actually, nvm was misreading the code. We are unwrapping the `Option` not the `Result>

**Review by thinkharderdev — APPROVED** [review]
LGTM

**Pull request merged**